File: 310.minimum-height-trees.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/*
 * @lc app=leetcode id=310 lang=java
 *
 * [310] Minimum Height Trees
 */

// @lc code=start
class Solution {

    public ArrayList<ArrayList<Integer>> makeAdjacencyList(int[][] pre, int num) {
        ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>>();

        for (int i = 0; i < num; i++) {
            adj.add(new ArrayList<>());
        }

        for (int[] course : pre) {
            adj.get(course[0]).add(course[1]);
            adj.get(course[1]).add(course[0]);

        }

        return adj;
    }

    public List<Integer> findMinHeightTrees(int n, int[][] edges) {

      
    }
}
// @lc code=end

File: 310.minimum-height-trees.py
#
# @lc app=leetcode id=310 lang=python3
#
# [310] Minimum Height Trees
#

# @lc code=start
from typing import List


class Solution:
    
      
        
    
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]
        adj = [[] for _ in range(n)]
        for edge in edges:
            adj[edge[0]].append(edge[1])
            adj[edge[1]].append(edge[0])
        leaves = [i for i in range(n) if len(adj[i]) == 1]
        while n > 2:
            n -= len(leaves)
            newLeaves = []
            for leaf in leaves:
                neighbor = adj[leaf].pop()
                adj[neighbor].remove(leaf)
                if len(adj[neighbor]) == 1:
                    newLeaves.append(neighbor)
            leaves = newLeaves
        return leaves
                
            
        

        
# @lc code=end

